// @ngneat/helipopper/esm2022/lib/tippy.service.mjs
import { Inject, Injectable, Injector } from '@angular/core';
import tippy from 'tippy.js';
import { isComponent, isTemplateRef } from '@ngneat/overview';
import { TIPPY_CONFIG, TIPPY_REF } from './tippy.types';
import { normalizeClassName, onlyTippyProps } from './utils';
import * as i0 from "@angular/core";
import * as i1 from "@ngneat/overview";
export class TippyService {
    constructor(globalConfig, view, injector) {
        this.globalConfig = globalConfig;
        this.view = view;
        this.injector = injector;
    }
    create(host, content, options = {}) {
        const variation = options.variation || this.globalConfig.defaultVariation;
        const config = {
            onShow: (instance) => {
                host.setAttribute('data-tippy-open', '');
                if (!instance.$viewOptions) {
                    instance.$viewOptions = {};
                    const injector = Injector.create({
                        providers: [
                            {
                                provide: TIPPY_REF,
                                useValue: instance,
                            },
                        ],
                        parent: options.injector || this.injector,
                    });
                    instance.$viewOptions.injector = injector;
                    if (isTemplateRef(content)) {
                        instance.$viewOptions.context = {
                            $implicit: instance.hide.bind(instance),
                            ...options.context,
                        };
                    }
                    else if (isComponent(content)) {
                        instance.context = options.context;
                        instance.data = options.data;
                    }
                }
                if (!instance.view) {
                    instance.view = this.view.createView(content, { ...options, ...instance.$viewOptions });
                }
                instance.setContent(instance.view.getElement());
                options?.onShow?.(instance);
            },
            onHidden: (instance) => {
                host.removeAttribute('data-tippy-open');
                if (!options.preserveView) {
                    instance.view.destroy();
                    instance.view = null;
                }
                options?.onHidden?.(instance);
            },
            ...onlyTippyProps(this.globalConfig),
            ...this.globalConfig.variations[variation],
            ...onlyTippyProps(options),
            onCreate: (instance) => {
                instance.popper.classList.add(`tippy-variation-${variation}`);
                if (options.className) {
                    for (const klass of normalizeClassName(options.className)) {
                        instance.popper.classList.add(klass);
                    }
                }
                this.globalConfig.onCreate?.(instance);
                options.onCreate?.(instance);
            },
        };
        return tippy(host, config);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.4", ngImport: i0, type: TippyService, deps: [{ token: TIPPY_CONFIG }, { token: i1.ViewService }, { token: i0.Injector }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.0.4", ngImport: i0, type: TippyService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.4", ngImport: i0, type: TippyService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TIPPY_CONFIG]
                }] }, { type: i1.ViewService }, { type: i0.Injector }] });
